### PR TITLE
Allow subclassing of RegExps

### PR DESCRIPTION
**Pull request by mikesamuel**

### CODE REVIEW

**Comment by ljharb** [comment]
:+1:

File: RegExp.make.js [hunk 1]
```diff
@@ -1065,12 +1070,26 @@ RegExp.make = (function () {
...
   }
 
-  return make.bind(null, '' /* No flags by default */);
+  return function(x, ...values) {
+    // RegExp.make can be called in several modes.
+    // 1. RegExp.make`...undifferentiated RegExp stuff...`
+    // 2. RegExp.make('gi')`....` to specify flags
+    // 3. RegExp.make.bind(RegExpSubClass)`...` with a this value that specifies
+    //    a different constructor.
+    const xType = typeof x;
```
Comment (bergus) [inline RegExp.make.js]: Just use `typeof x` twice. Should be shorter and easier to read; also engines optimise this better when `typeof` is a direct operand of `==`/`===`.

File: tests.js [hunk 2]
```diff
@@ -9,9 +9,47 @@
     return;
   }
 
-  function qpair(template, ...values) {
-    return [template, values];
+  // Set up a RegExp subclass so that we can test subclass creation.
+  function SubRegExp(source, flags) {
+    RegExp.call(this, source, flags);
```
Comment (bergus) [inline tests.js]: Not sure whether this makes sense at all while ES6 subclassing is not supported. Admittedly, `RegExp.call(this, source, flags);` throws the errors we would expect, but it doesn't modify `this` (neither in ES5 nor ES6).

Comment (ljharb) [inline tests.js]: I believe you'd have to do `var re = new RegExp(source, flags); Object.setPrototypeOf(re, SubRegExp.prototype); return re;` as well as `Object.setPrototypeOf(SubRegExp, RegExp);` ([example](https://github.com/paulmillr/es6-shim/blob/master/test/collections.js#L131-L139))

Comment (bergus) [inline tests.js]: Yeah, that sounds like a reasonable shim for ES6 subclassing.

File: RegExp.make.js (hunk 1 above)
Comment (mikesamuel) [inline RegExp.make.js]: Done

File: tests.js (hunk 2 above)
Comment (mikesamuel) [inline tests.js]: I did the first.  Instead of doing  `Object.setPrototypeOf(SubRegExp, RegExp);` I modified the example ljharb pointed to, specifically https://github.com/paulmillr/es6-shim/blob/master/test/collections.js#L137

**Comment by mikesamuel** [comment]
@bergus, are you ok with patch as it stands?

File: RegExp.make.js [hunk 3]
```diff
@@ -306,7 +306,7 @@ RegExp.make = (function () {
    *    a start context, and RegExp source, and returns an end context.
    */
   function parseRegExpSource(eventHandler) {
-    const {
+    var {
```
Comment (ljharb) [inline RegExp.make.js:309]: why is this changed to `var`?

File: tests.js [hunk 4]
```diff
@@ -9,9 +9,50 @@
...
+          ctor: ctor,
+          flags: flags,
+          template: template,
+          values: values.slice()
+        };
+      } else {
+        var makerCtor = ctor;
+        var makerFlags = flags;
+        const args = [x, ...values];
+        for (var i = 0, n = args.length; i < n; ++i) {
```
Comment (ljharb) [inline tests.js:45]: Why is this a loop rather than a `args.forEach`?

File: tests.js [hunk 5]
```diff
@@ -9,9 +9,50 @@
...
+          const arg = args[i];
+          switch (typeof arg) {
+          case 'function': makerCtor = arg; break;
+          case 'string':   makerFlags = arg; break;
+          }
```
Comment (ljharb) [inline tests.js:50]: rather than squishing multiple statements onto one line, could you use if/else instead of a switch statement? `switch` is prone to bugs, and is more verbose and less maintainable than other options.

File: tests.js (hunk 4 above)
Comment (mikesamuel) [inline tests.js:45]: Changed to use forEach.

File: tests.js (hunk 5 above)
Comment (mikesamuel) [inline tests.js:50]: Done and the terminal else throws.

File: RegExp.make.js (hunk 3 above)
Comment (ljharb) [inline RegExp.make.js:309]: still not sure why this is changed to `var`

File: RegExp.make.js [hunk 6]
```diff
@@ -1065,12 +1076,25 @@ RegExp.make = (function () {
       pattern += rawLiteralPart;
       addTemplateGroups(i+1);
     }
-    const output = new RegExp(pattern, flags);
+    var output = new ctor(pattern, flags);
```
Comment (ljharb) [inline RegExp.make.js:1079]: or this one

File: RegExp.make.js (hunk 3 above)
Comment (mikesamuel) [inline RegExp.make.js:309]: This is the same limitation of closure compiler.  It types `const`s in a separate pass from `var`s.
